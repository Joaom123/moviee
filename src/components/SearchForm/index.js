import React, {Component} from "react";

class SearchForm extends Component {
    constructor(props) {
        super(props);

        this.state = {
            valor: '',
        };

        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);
    }

    handleChange(event) {
        this.setState({
            valor: event.target.value
        });
    }

    async handleSubmit(event) {
        event.preventDefault();
        await this.props.onSubmit(this.state.valor);
    }

    render() {
        return (
            <div className="row">
                <form className="col s12" action="">
                    <div className="row">
                        <div className="input-field col s6">
                            <input className="validate" type="text" id="nomeFilme" value={this.state.valor} onChange={this.handleChange}/>
                            <label htmlFor="nomeFilme">Nome do Filme</label>
                        </div>
                        <input type="submit" value="Pesquisar"
                               onClick={(event) => this.handleSubmit(event)}/>
                    </div>
                </form>
            </div>
        );
    }
}

export default SearchForm;