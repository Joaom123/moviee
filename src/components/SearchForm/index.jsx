import React from "react";

function SearchForm({searchValue, handleChange}) {
    const onSubmit = (event) => event.preventDefault();

    return(
        <section className="row">
            <form className="col s12" onSubmit={onSubmit}>
                <div className="row">
                    <div className="input-field col s12 m6 offset-m3">
                        <input
                            className="validate"
                            id="movieName"
                            name="movieName"
                            type="text"
                            value={searchValue}
                            onChange={handleChange}
                        />
                        <label htmlFor="movieName">Title</label>
                    </div>
                </div>
            </form>
        </section>
    );
}

export default SearchForm;