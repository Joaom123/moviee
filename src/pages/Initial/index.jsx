import React, {Component} from 'react';
import {ModalMovie, Movies, Pagination, SearchForm, Snackbar} from "../../components";
import {apiService} from "../../services";

class Initial extends Component {
    constructor(props) {
        super(props);

        this.state = {
            searchValue: "",
            page: 1,
            movies: [],
            movie: {},
            totalResults: 0,
            toggleModal: false,
            errorMessage: "",
            isActive: false
        };

        this.handleChangeOfSearchValue = this.handleChangeOfSearchValue.bind(this);
        this.handleChangeOfPage = this.handleChangeOfPage.bind(this);
        this.onCardClick = this.onCardClick.bind(this);
    }

    onCloseModal = () => this.setState({toggleModal: false, movie: {}});

    doesGetMoviesListIfValidSearchValue = () => {
        if (this.isSearchValueValid())
            this.getMoviesListAndSetMovies();
    }

    doesGetMoviesListWithTimeout = () => {
        clearTimeout(this.timer);
        this.timer = setTimeout(
            () => this.doesGetMoviesListIfValidSearchValue(),
            800
        );
    }

    isSearchValueNotEmpty = () => this.state.searchValue !== "";

    isSearchValueLongerThan3Characters = () => this.state.searchValue.length >= 3;

    isSearchValueValid = () => this.isSearchValueNotEmpty() && this.isSearchValueLongerThan3Characters();

    getMoviesListAndSetMovies = () => {
        const {searchValue, page} = this.state;
        apiService
            .getMoviesList({searchValue, page})
            .then(response => {
                const {Response, Search, totalResults, Error} = response.data;

                if (Response === 'True') {
                    this.setState({
                        movies: Search,
                        totalResults
                    });
                } else {
                    this.setState({
                        errorMessage: Error,
                        isActive: true,
                        movies: [],
                        page: 1,
                        totalResults: 0
                    });
                }
            })
            .catch(error => {
                console.log(error);
            })
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        const {searchValue, page, isActive} = this.state;

        if(searchValue !== prevState.searchValue || page !== prevState.page)
            this.doesGetMoviesListWithTimeout();

        if(isActive === true && isActive !== prevState.isActive) {
            setTimeout(
                () => this.setState({isActive: false}),
                3000
            );

        }
    }

    handleChangeOfSearchValue(event) {
        let searchValue = event.target.value;

        if (searchValue === " ")
            return ;

        this.setState({searchValue});
    }

    handleChangeOfPage(e, page) {
        e.preventDefault();
        this.setState({page});
    }

    onCardClick(imdbId) {
        apiService
            .getMovieByImdbId({imdbId})
            .then(response => {
                this.setState({
                    movie: response.data,
                    toggleModal: true
                });
            });
    }

    render() {
        const {searchValue, movies, totalResults, page, movie, toggleModal, errorMessage, isActive} = this.state;

        return (
            <main className="container">
                <SearchForm
                    searchValue={searchValue}
                    handleChange={this.handleChangeOfSearchValue}
                />
                <Movies
                    movies={movies}
                    onCardClick={this.onCardClick}
                />
                <Pagination
                    totalResults={totalResults}
                    page={page}
                    handleChangeOfPage={this.handleChangeOfPage}
                />
                <ModalMovie
                    toggleModal={toggleModal}
                    onCloseModal={this.onCloseModal}
                    movie={movie}
                />
                <Snackbar
                    isActive={isActive}
                    message={errorMessage}
                />
            </main>
        );
    }
}

export default Initial;
