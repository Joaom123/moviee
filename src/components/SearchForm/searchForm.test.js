import React from "react";
import {cleanup, render} from "@testing-library/react";
import Initial from "../../pages/Initial";
import SearchForm from "./index";

afterEach(cleanup);

describe("SearchForm", () => {
    it('snapshot', function () {
        const {asFragment} = render(<SearchForm />);
        expect(asFragment(<SearchForm />)).toMatchSnapshot();
    });

    it('should render search value', function () {
        const searchValue = "test";
        const {getByTestId} = render(<SearchForm handleChange={() => {}} searchValue={searchValue} />);
        expect(getByTestId("searchMovieInput").value).toBe(searchValue);
    });
});