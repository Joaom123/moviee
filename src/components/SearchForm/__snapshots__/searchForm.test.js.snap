// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm snapshot 1`] = `
<DocumentFragment>
  <section
    class="row"
  >
    <form
      class="col s12"
    >
      <div
        class="row"
      >
        <div
          class="input-field col s12 m6 offset-m3"
        >
          <input
            class="validate"
            data-testid="searchMovieInput"
            id="movieName"
            name="movieName"
            type="text"
            value=""
          />
          <label
            for="movieName"
          >
            Title
          </label>
        </div>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
